src/vmm/src/lib.rs

```diff
@@ -1763,6 +1795,90 @@ mod tests {
                 }
             }
         }
+
+        #[cfg(target_arch = "x86_64")]
+        fn config_vcpu(&self, vcpu_fd: &mut VcpuFd, guest_addr: GuestAddress) {
+            // x86_64 specific registry setup.
+            let mut vcpu_sregs = vcpu_fd.get_sregs().unwrap();
+            vcpu_sregs.cs.base = 0;
+            vcpu_sregs.cs.selector = 0;
+            vcpu_fd.set_sregs(&vcpu_sregs).unwrap();
+
+            let mut vcpu_regs = vcpu_fd.get_regs().unwrap();
+            // Set the Instruction Pointer to the guest address where we loaded the code.
+            vcpu_regs.rip = guest_addr.0;
+            vcpu_regs.rax = 2;
+            vcpu_regs.rbx = 3;
+            vcpu_regs.rflags = 2;
+            vcpu_fd.set_regs(&vcpu_regs).unwrap();
+        }
+
+        #[cfg(target_arch = "aarch64")]
+        fn config_vcpu(&self, vcpu_fd: &mut VcpuFd, guest_addr: GuestAddress) {
+            // aarch64 specific registry setup.
+            let mut kvi = kvm_bindings::kvm_vcpu_init::default();
+            self.vm.fd().get_preferred_target(&mut kvi).unwrap();
+            vcpu_fd.vcpu_init(&kvi).unwrap();
+
+            let core_reg_base: u64 = 0x6030_0000_0010_0000;
+            let mem_size = self.vm_config.mem_size_mib.unwrap() << 20;
+            let mmio_addr: u64 = guest_addr.0 + mem_size as u64;
+            vcpu_fd
+                .set_one_reg(core_reg_base + 2 * 32, guest_addr.0)
+                .unwrap(); // set PC
+            vcpu_fd
+                .set_one_reg(core_reg_base + 2 * 0, mmio_addr)
+                .unwrap(); // set X0
+        }
+
+        // Prepare a VM with 1 MB memory and 1 vCPU.
+        // Dirty a page and run the VM.
+        // Return after the first VM exit.
+        fn setup_with_code_and_run(&mut self) {
```

Comment:
I can't find the new PR so I'm writing here just to make sure I don't forget.

This logic seems quite complicated. Could we test this by loading the `test_elf.bin` from the `kernel` crate and expect for a certain dirty map ? Or create a special elf that dirties only one page ?

Reply:
If we load a full blown kernel, there's no way of knowing for sure how many pages it will dirty; and creating a custom kernel seems more complicated to me than loading a handful of asm (I also find it more transparent to view the asm itself). Do you think we should add more comments to explain the instructions?

Reply:
Sounds good. I'll restructure the code for a more natural flow.
What kind of write do you mean? An MMIO is one option - it triggers a VM exit and validates the MMIO flow, but doesn't change anything in the dirty log - or an actual write in guest memory? The latter will dirty one more page, so we'd need to change the test passing condition (one page dirtied on `aarch64` and two on `x86_64`).

Reply:
I meant an actual write in the guest memory. It would be nice to have the same logic for `ARM` and `x86_64`. So a write in the guest memory + a MMIO write. I thought this would dirty only one page.

Also do we actually need to generate a HALT or MMIO write ? Can't we get the dirty log while `KVM_RUN` is executing ?